File: examples/training/05-observability-guardrails/step1-child-loggers.ts
import { createAgent } from '@agentuity/runtime';

export default createAgent('child-logger-demo', {
  handler: async (ctx, input) => {
    const data = input as { userId?: string; requestId?: string; query?: string };
    const userId = data.userId || 'unknown';
    const requestId = data.requestId || ctx.runId;

    // Create child logger with context
    // All logs from this child logger will include these fields
    const logger = ctx.logger.child({
      requestId,
      userId
    });

    // Log processing start
    logger.info('Processing request started');

    // Simulate some processing steps
    logger.debug('Validating input data');

    if (!data.query) {
      logger.warn('Missing query field');
      return {
        message: 'Query is required',
        processedAt: new Date().toISOString()
      };
    }

    logger.debug('Processing query', { queryLength: data.query.length });

    // Simulate successful processing
    const result = {
      message: 'Request processed successfully',
      query: data.query,
      processedAt: new Date().toISOString()
    };

    logger.info('Processing completed successfully');

    return result;
  }
});

File: examples/training/05-observability-guardrails/step1-child-loggers.py
from agentuity import AgentRequest, AgentResponse, AgentContext
from datetime import datetime


async def run(request: AgentRequest, response: AgentResponse, ctx: AgentContext):
    """Demonstrates logging with context for easier debugging."""
    data = await request.data.json()
    user_id = data.get("userId", "unknown")
    request_id = data.get("requestId", ctx.run_id)

    # Python SDK child logging coming soon
    # For now, include context manually in structured logs
    log_context = {"requestId": request_id, "userId": user_id}

    # Log processing start
    ctx.logger.info("Processing request started", log_context)

    # Simulate some processing steps
    ctx.logger.debug("Validating input data", log_context)

    query = data.get("query")
    if not query:
        ctx.logger.warning("Missing query field", log_context)
        return response.json({"error": "Query is required"})

    ctx.logger.debug("Processing query", {**log_context, "queryLength": len(query)})

    # Simulate successful processing
    result = {
        "message": "Request processed successfully",
        "query": query,
        "processedAt": datetime.now().isoformat(),
    }

    ctx.logger.info("Processing completed successfully", log_context)

    return response.json(result)
